pilot/pkg/status/resourcelock.go

```diff
@@ -14,39 +14,109 @@
 
 package status
 
-import "sync"
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+)
+import "k8s.io/apimachinery/pkg/runtime/schema"
 
-type ResourceLock struct {
+type ResourceMutex struct {
 	masterLock sync.RWMutex
-	listing    map[interface{}]*sync.Mutex
+	cache map[lockResource]*cacheEntry
 }
 
-func (r *ResourceLock) Lock(i interface{}) {
-	r.retrieveMutex(i).Lock()
+type cacheEntry struct {
+	// the runlock ensures only one routine is writing status for a given resource at a time
+	runLock sync.Mutex
+	// the cachelock protects writes and reads to the two cache values
+	cacheLock sync.Mutex
+	// the cacheVale represents the latest version of the resource, including ResourceVersion
+	cacheVal *Resource
+	// the cacheProgress represents the latest version of the Progress
+	cacheProgress *Progress
+	countLock     sync.Mutex
+	// track how many routines are running for this resource.  always <= 2
+	routineCount  int32
+	// track if this resource has been deleted, and if so, stop writing updates for it
+	deleted       bool
 }
 
-func (r *ResourceLock) Unlock(i interface{}) {
-	r.retrieveMutex(i).Unlock()
+func (rl *ResourceMutex) Delete(r Resource) {
+	f := rl.retrieveEntry(convert(r))
+	f.cacheLock.Lock()
+	defer f.cacheLock.Unlock()
+	f.deleted = true
+}
+
+// OncePerResource will run the target function at most once per resource, with up to one waiting routine per resource
+// we only want one update per resource running concurrently to prevent overwhelming the API server, but we also
+// don't want to stack up goroutines when the API server is overwhelmed.  This limits us to 2*resourceCount goroutines
+// while ensuring that when a goroutine begins, it gets the latest status values for the resource.
+func (rl *ResourceMutex) OncePerResource(ctx context.Context, r Resource, p Progress, target func(ctx context.Context, config Resource, state Progress)) {
+	lr := convert(r)
+	f := rl.retrieveEntry(lr)
+	// every call updates the cache, so whenever the standby routine fires, it runs with the latest values
+	f.cacheLock.Lock()
+	f.cacheVal = &r
+	f.cacheProgress = &p
+	f.cacheLock.Unlock()
+	// we only fire off goroutine if routincount < 2, so we can have a currently running routine and a standby routine
+	if atomic.LoadInt32(&f.routineCount) < 2 {
+		f.countLock.Lock()
+		if atomic.LoadInt32(&f.routineCount) < 2 {
+			// double check routinecount before incrementing to avoid race
+			atomic.AddInt32(&f.routineCount, 1)
+			f.countLock.Unlock()
+			go func() {
+				f.runLock.Lock()
```

Comment:
I think this can be changed to have a single goroutine and less locks possibly. A call to the function will refresh the "pending" value. If there isn't anything running, it will also start running. If there is something running, do nothing. The running goroutine will continue to run until there is nothing pending.

Something like
```go
func OncePerResource() {
	f.cacheLock.Lock()
	if f.running {
		// currently running goroutine will pick up the change on completion
		f.cacheVal = &r
		f.cacheLock.Unlock()
		return
	}

	val := f.cacheVal
	f.cacheVal = nil
	f.running = true
	f.cacheLock.Unlock()
	go func() {
		for {
			target(ctx, val)
			f.cacheLock.Lock()
			if f.cacheVal != nil {
				val = f.cacheVal
				f.cacheVal = nil
			} else {
				f.running = false
				f.cacheLock.Unlock()
				return
			f.cacheLock.Unlock()
		}
	}
	}
```

Reply:
The idea is to have one running routine and one waiting routine to start immediately after the running one finishes.  I'll see if I can't clean this up as it does look scary.

Reply:
> The idea is to have one running routine and one waiting routine to start immediately after the running one finishes. I'll see if I can't clean this up as it does look scary.

Right, my suggestion is to do it with a single goroutine; the end result will be the same. There is no need to have a second waiting goroutine, we can just loop in the running goroutine if there is more work to be done

Reply:
Or do we even want 1 go routine per resource? For example, if a user modifies 1k resources at in short succession, we queue up 2k (or 1k if we change to above) goroutines, which all write to the apiserver at once and get heavily throttled. Is there any benefits over just having some standard work queue with some bounded parallelism?